write-fonts/src/tables/gsub/builders.rs

```diff
@@ -2,10 +2,95 @@
 
 use std::{collections::BTreeMap, convert::TryFrom};
 
-use types::GlyphId16;
+use types::{FixedSize, GlyphId16, Offset16};
 
-use crate::tables::{layout::builders::Builder, variations::ivs_builder::VariationStoreBuilder};
+use crate::{
+    tables::{
+        layout::{builders::Builder, CoverageTable},
+        variations::ivs_builder::VariationStoreBuilder,
+    },
+    FontWrite,
+};
 
+/// Helper type for splitting layout subtables
+// NOTE: currently just used for ligature substitution, but hopefully can be
+// reused for other lookups as needed?
+#[derive(Clone, Debug)]
+struct TableSplitter<T: SplitTable> {
+    finished: Vec<T>,
+    current_coverage: Vec<GlyphId16>,
+    current_items: Vec<T::Component>,
+    current_size: usize,
+}
+
+/// A trait for splitting layout subtables.
+trait SplitTable {
+    /// The component item of this table.
+    type Component;
+
+    /// The (maximum) number of bytes required to store this item.
+    ///
+    /// This should include only the costs of the item, not the cost of adding
+    /// to the coverage table.
+    ///
+    /// 'Maximum' because this does not account for possible size savings from
+    /// deduplication of identical objects, for instance.
+    fn size_for_item(item: &Self::Component) -> usize;
+    /// The starting size of a new table that will contain this item.
+    ///
+    /// This does not include the size of the item itself! The item is provided
+    /// because sometimes the table contains fields derived from its members.
+    fn initial_size_for_item(item: &Self::Component) -> usize;
+    fn instantiate(coverage: CoverageTable, items: Vec<Self::Component>) -> Self;
+}
+
+impl<T: SplitTable + FontWrite> TableSplitter<T> {
+    const MAX_TABLE_SIZE: usize = u16::MAX as usize;
+
+    fn new() -> Self {
+        Self {
+            finished: Vec::new(),
+            current_coverage: Vec::new(),
+            current_items: Vec::new(),
+            current_size: 0,
+        }
+    }
+
+    fn add(&mut self, gid: GlyphId16, item: T::Component) {
+        let item_size = T::size_for_item(&item);
+        if item_size + self.current_size > Self::MAX_TABLE_SIZE {
+            let current_len = self.current_coverage.len();
+            self.finish_current();
+            let type_ = self.finished.last().unwrap().table_type();
+            log::info!("adding split in {type_} at {current_len}");
+        }
+
+        if self.current_size == 0 {
+            self.finish_current();
```

Comment:
I would assume that if `self.current_size` is 0 then `finish_current()` would have nothing to do. That _seems_ to be the case based on my reading of the code but I’m not sure if I’m missing something.

Reply:
oh yea this is an extra statement; it's a noop because we've already called `finish_current` above, and `finish_current` is a noop if we haven't added any items yet.